### PR TITLE
More WebDataset docs

### PR DESCRIPTION
**Pull request by lhoestq**

### CODE REVIEW

File: docs/hub/datasets-webdataset.md [hunk 1]
```diff
@@ -1,20 +1,74 @@
 # WebDataset
 
-[WebDataset](https://github.com/webdataset/webdataset) is a library to write I/O pipelines for large datasets. 
-Since it supports streaming data using HTTP, you can use the Hugging Face data files URLs to stream a dataset in WebDataset format:
+[WebDataset](https://github.com/webdataset/webdataset) is a library to write I/O pipelines for large datasets.
+It is especially useful to stream large scale datasets in WebDataset format to a DataLoader.
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:4]: I think a few words about why its useful would be helpful

```suggestion
Its sequential I/O and sharding features make it especially useful for streaming large-scale datasets to a DataLoader.
```

File: docs/hub/datasets-webdataset.md [hunk 2]
```diff
@@ -1,20 +1,74 @@
...
+
+## The WebDataset format
+
+A WebDataset file is a TAR archive containing a series of data files.
+All the successive data files with the same prefix are considered to be part of the same example (e.g. an image/audio file and its label or metadata):
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:9]:
```suggestion
All successive data files with the same prefix are considered to be part of the same example (e.g., an image/audio file and its label or metadata):
```

File: docs/hub/datasets-webdataset.md [hunk 3]
```diff
@@ -1,20 +1,74 @@
...
+
+<div class="flex justify-center">
+<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-webdataset-format.png"/>
+<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-webdataset-format-dark.png"/>
+</div>
+
+A large scale WebDataset is made of many files called shards, where each shard is a TAR archive.
+Each shard is often ~1GB but the full dataset can be multiple terabytes !
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:17]:
```suggestion
Each shard is often ~1GB but the full dataset can be multiple terabytes!
```

File: docs/hub/datasets-webdataset.md [hunk 4]
```diff
@@ -1,20 +1,74 @@
...
+
+## Stream using the WebDataset library
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:19]:
```suggestion
## Streaming
```

File: docs/hub/datasets-webdataset.md [hunk 5]
```diff
@@ -1,20 +1,74 @@
...
+
+Streaming TAR archives is fast because it reads contiguous chunks of data.
+It can be orders of magnitude faster than reading separate data files one by one.
+
+Therefore WebDataset streaming achieves top speed performance both when reading from disk and from a cloud storage, which makes it an ideal solution to feed a DataLoader:
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:24]:
```suggestion
WebDataset streaming offers high-speed performance both when reading from disk and from cloud storage, which makes it an ideal format to feed to a DataLoader:
```

File: docs/hub/datasets-webdataset.md [hunk 6]
```diff
@@ -1,20 +1,74 @@
...
+For example here is how to stream the [timm/imagenet-12k-wds](https://huggingface.co/datasets/timm/imagenet-12k-wds) dataset directly from Hugging Face:
 
 First you need to [Login with your Hugging Face account](../huggingface_hub/quick-start#login), for example using:
 
 ```
 huggingface-cli login
 ```
 
-And then you can stream Hugging Face datasets in WebDataset. Here is an example of how to load [timm/imagenet-12k-wds](https://huggingface.co/datasets/timm/imagenet-12k-wds):
+And then you can stream the Hugging Face dataset in WebDataset:
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:39]:
```suggestion
And then you can stream the dataset with WebDataset:
```

File: docs/hub/datasets-webdataset.md [hunk 7]
```diff
@@ -1,20 +1,74 @@
...
->>> dataset = wds.WebDataset(f"pipe:curl -s -L https://huggingface.co/datasets/timm/imagenet-12k-wds/resolve/main/imagenet12k-train-{{0000..1023}}.tar -H 'Authorization:Bearer {hf_token}'")
+>>> url = "https://huggingface.co/datasets/timm/imagenet-12k-wds/resolve/main/imagenet12k-train-{{0000..1023}}.tar"
+>>> url = f"pipe:curl -s -L {url} -H 'Authorization:Bearer {hf_token}'"
+>>> dataset = wds.WebDataset(url).decode()
+>>> dataloader = DataLoader(dataset, batch_size=64, num_workers=4)
+```
+
+## Shuffle
+
+Generally, datasets in WebDataset formats are already shuffled and ready to feed a DataLoader.
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:55]:
```suggestion
Generally, datasets in WebDataset formats are already shuffled and ready to feed to a DataLoader.
```

File: docs/hub/datasets-webdataset.md [hunk 8]
```diff
@@ -1,20 +1,74 @@
...
+But it is still possible to re-shuffle the data using the WebDataset approximate shuffling.
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:56]:
```suggestion
But you can still reshuffle the data with WebDataset's approximate shuffling.
```

File: docs/hub/datasets-webdataset.md [hunk 9]
```diff
@@ -1,20 +1,74 @@
...
+
+In addition to shuffling the list of shards, WebDataset uses a buffer to shuffle a dataset without any speed cost:
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:58]:
```suggestion
In addition to shuffling the list of shards, WebDataset uses a buffer to shuffle a dataset without any cost to speed:
```

File: docs/hub/datasets-webdataset.md [hunk 10]
```diff
@@ -1,20 +1,74 @@
...
+
+<div class="flex justify-center">
+<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-webdataset-shuffle.png"/>
+<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-webdataset-shuffle-dark.png"/>
+</div>
+
+Here is how to shuffle the shards list and use a shuffle buffer to shuffle a WebDataset:
```
Comment (stevhliu) [inline docs/hub/datasets-webdataset.md:65]:
```suggestion
To shuffle a list of sharded files and randomly sample from the buffer:
```

**Review by stevhliu — APPROVED** [review]
Awesome work! 👍  Only comments are minor nits and suggestions :)

**Review by rwightman — APPROVED** [review]

**Comment by rwightman** [comment]
@lhoestq what's there is fine, wds examples are simple, appear correct.

Two comments that can be addressed in follow up additions
1. Why no example using datasets to stream the wds shards? It does work, but as discussed needs the http streaming fix to be similar to webdataset + curl, and needs distributed improvements
2. The webdataset example covers the in-order most basic case, and simple shard + sample shuffling cases. Distributed training, epoch based training, shard resampling with replacement use cases could also be documented at some point, though they get quite a bit more complex to cover the nuances as you can see in timm or open_clip code...

**Comment by lhoestq** [comment]
Thanks !

> Why no example using datasets to stream the wds shards? It does work, but as discussed needs the http streaming fix to be similar to webdataset + curl, and needs distributed improvements

it will be in the `datasets` docs :)

> The webdataset example covers the in-order most basic case, and simple shard + sample shuffling cases. Distributed training, epoch based training, shard resampling with replacement use cases could also be documented at some point, though they get quite a bit more complex to cover the nuances as you can see in timm or open_clip code...

Definitely ! That's the next step ^^ Btw if you have some good references on this that would be helpful

**Comment by lhoestq** [comment]
Merging for now, but lmk if you have more comments
